[UIMA-6243] JCas returns Annotation when asked for a specific subtype

reckart opened the pull request:
- Implemented unit test which demonstrates the issue

Jira issue: https://issues.apache.org/jira/browse/UIMA-6243

mischor commented on uimaj-core/pom.xml:
```diff
@@ -78,6 +78,11 @@
 			<artifactId>assertj-core</artifactId>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
```
This dependency should have a <scope> of test, right?

reckart replied:
I only added test code - so yes

mischor commented on uimaj-core/src/test/java/org/apache/uima/cas/test/JCasClassLoaderTest.java:
```diff
@@ -0,0 +1,292 @@
+
+    public IsolatingClassloader redefining(String... patterns)
+    {
+      hideClassesPatterns.addAll(asList(patterns));
```
should this be adding these patterns to "redefineClassesPatterns", not "hideClassesPatterns"?

reckart replied:
Indes

mischor replied:
OK, see long comment in the Jira.  I'm thinking the test case has to be rewritten a bit?

reckart commented:
Thanks for the check. I have updated the test. 

Now, one test fails in UIMAv2 (with `Annotation` where `Token` would be expected) and the other two fail in UIMAv3 (with the `Token` cannot be cast to `Token` exception).

I have also updated the test with JavaDoc trying to explain the setup and the expected results.

reckart commented:
Mind, the test is still not quite perfect. There should actually be some asserts actually checking that the expected JCas wrappers are being used instead of just implicitly assuming that they are using if the checks don't fail with class-cast exceptions.

reckart commented:
@mischor ok, now we have proper asserts

mischor commented:
Jenkins, please test this

reckart commented:
Btw. here is a list of the available commands: https://plugins.jenkins.io/ghprb/ - not sure if your word order matched and triggered a build

mischor commented:
@reckart can you see if this works (in uima v2) for you?

reckart commented:
@mischor I'll try next week. The setup to reproduce this "in vivo" is not something I deal with every day (involving Docker, OSGI and whatnot)... that's why accompanied with a lot of head-scratching I implemented the unit test ;)

reckart commented:
@mischor looks good to me - unit tests are also all passing. I added another unit test which uses a custom  classpath instead of a custom classloader - also works. So I'll merge this one.
